scripts/alert_emails/send_emails.ts

```diff
@@ -138,99 +46,104 @@ async function setLastSnapshotNumber(
-                  `Rate limit remaining: ${rateLimitRemaining}. Sleeping ${rateLimitReset} seconds.`,
-                );
-                await delay(rateLimitReset * 1000);
-              }
-              // Comment out this code if you are developing locally
-              await db
-                .collection(
-                  `snapshots/${currentSnapshot}/locations/${fips}/emails/`,
-                )
-                .doc(doc.id)
-                .set({
-                  sentAt: admin.firestore.FieldValue.serverTimestamp(),
-                });
-            })
-            .catch(async (err: CampaignMonitorError) => {
-              // if the email is invalid, move the document to the invalid collection
-              if (
-                err.Code === CM_INVALID_EMAIL_ERROR_CODE &&
-                err.Message === CM_INVALID_EMAIL_MESSAGE
-              ) {
-                invalidEmailCount += 1;
-                const currentData = (
-                  await db.collection('alerts-subscriptions').doc(doc.id).get()
-                ).data();
-                if (currentData) {
-                  await db
-                    .collection('invalid-alerts-subscriptions')
-                    .doc(doc.id)
-                    .set(currentData);
-                  await db
-                    .collection('alerts-subscriptions')
-                    .doc(doc.id)
-                    .delete();
-                }
-              } else {
-                errorCount += 1;
-              }
-            });
-        }
-      });
+      .get();
+    return querySnapshot.docs.map(emailDoc => emailDoc.id);
   }
+
+  async function onInvalidEmail(email: string) {
+    invalidEmailCount += 1;
+    const querySnapshot = await db
+      .collection('alerts-subscriptions')
+      .doc(email)
+      .get();
+    const currentData = querySnapshot.data();
+    if (currentData) {
+      await db
+        .collection('invalid-alerts-subscriptions')
+        .doc(email)
+        .set(currentData);
+      await db.collection('alerts-subscriptions').doc(email).delete();
+    }
+  }
+
+  async function onEmailSent(email: string, fips: string) {
+    emailSent += 1;
+    uniqueEmailAddress[email] = (uniqueEmailAddress[email] || 0) + 1;
+    locationsWithEmails[fips] = (locationsWithEmails[fips] || 0) + 1;
+    // comment out this code if you are developing locally
+    return db
+      .collection(`snapshots/${currentSnapshot}/locations/${fips}/emails/`)
+      .doc(email)
+      .set({ sentAt: admin.firestore.FieldValue.serverTimestamp() });
+  }
+
+  async function sendAlertEmail(email: string, fips: string) {
+    const locationAlert = alertsByLocation[fips];
+    const sendData = generateAlertEmailData(email, locationAlert);
+    if (dryRun) {
+      await onEmailSent(email, fips);
+      return;
+    }
+
+    try {
+      const response = await createSend.sendClassicEmail(sendData);
+      await onEmailSent(email, fips);
+      return response;
```

Comment:
Hrm...  what is this response exactly and why do we return it (I can't tell if it's used)?

In general it's good TypeScript practice to define the types of any return values on your functions.  I think there's a tslint rule we could enable that would enforce it. 😁

Reply:
I'm adding a type for the `sendClassicEmail` method (I was missing the body of the response), but here we don't really need a specific response, just to return so the promise resolves.